internal/pkg/server/server.go

```diff
@@ -99,32 +101,46 @@ func (s *Server) streamImages(c echo.Context, pullAuth auth.Authenticator, image
 		return c.NoContent(http.StatusBadRequest)
 	}
 
-	tar, err := s.pullAndSaveImages(c.Request().Context(), pullAuth, images)
+	tar, err := s.pullAndSaveImages(c.Request().Context(), c, pullAuth, images)
 	if err != nil {
-		return err
+		return echo.ErrForbidden
 	}
 	defer tar.Close()
 
 	c.Response().Header().Set(echo.HeaderContentDisposition, `attachment; filename="images.tar"`)
 	return c.Stream(http.StatusOK, "application/x-tar", tar)
 }
 
-func (s *Server) pullAndSaveImages(ctx context.Context, authn auth.Authenticator, images []string) (io.ReadCloser, error) {
+func (s *Server) pullAndSaveImages(ctx context.Context, c echo.Context, authn auth.Authenticator, images []string) (io.ReadCloser, error) {
+	var wg sync.WaitGroup
+	wg.Add(len(images))
+
+	g, _ := errgroup.WithContext(ctx)
+
 	for _, img := range images {
-		encodedAuth, err := auth.RegistryAuthFor(authn, img)
-		if err != nil {
-			return nil, err
-		}
-		rc, err := s.DockerClient.ImagePull(ctx, img, types.ImagePullOptions{
-			RegistryAuth: encodedAuth,
+		img := img
+		g.Go(func() error {
+			encodedAuth, err := auth.RegistryAuthFor(authn, img)
+			if err != nil {
+				return err
+			}
+			rc, err := s.DockerClient.ImagePull(ctx, img, types.ImagePullOptions{
+				RegistryAuth: encodedAuth,
+			})
+			if err != nil {
+				return err
+			}
+			defer rc.Close()
+			io.Copy(os.Stdout, rc)
+			defer wg.Done()
```

Comment:
This WaitGroup will never resolve if an error happens before reaching this defer statement. Usually the `defer wg.Done()` should be placed at the top of such asynchronous functions to mitigate these problems. In this case the WaitGroup can be fully removed since sync.Errgroup handles the waiting for completion.